diff -Nru gc6.2.orig/configure.in gc6.2/configure.in
--- gc6.2.orig/configure.in	2003-06-21 06:57:18.000000000 +0200
+++ gc6.2/configure.in	2003-06-25 14:02:29.000000000 +0200
@@ -185,12 +185,10 @@
 case "$host" in
   *-*-darwin*) ;;
   *) 
-    AC_CHECK_LIB(dl, dlopen, EXTRA_TEST_LIBS="$EXTRA_TEST_LIBS -ldl")
+    AC_CHECK_LIB(dl, dlopen, LIBS="$LIBS -ldl")
     ;;
 esac
 
-AC_SUBST(EXTRA_TEST_LIBS)
-
 target_all=libgc.la
 AC_SUBST(target_all)
 
