--- gc-7.0/configure.ac.orig	2008-01-28 13:43:25.000000000 +0000
+++ gc-7.0/configure.ac	2008-01-28 13:44:30.000000000 +0000
@@ -392,6 +392,9 @@
     dnl This is now redundant, but it is also important for incremental GC
     dnl performance under Irix.
     ;;
+ sparc*-*-linux*) 
+    machdep="mach_dep.lo sparc_mach_dep.lo"
+    ;; 
  sparc-*-netbsd*)
     machdep="mach_dep.lo sparc_netbsd_mach_dep.lo"
     ;;
--- gc-7.0/mach_dep.c.orig	2008-01-30 14:46:14.000000000 +0000
+++ gc-7.0/mach_dep.c	2008-01-30 14:48:51.000000000 +0000
@@ -174,7 +174,7 @@
 
 #   if defined(HAVE_PUSH_REGS)
       GC_push_regs();
-#   elif defined(UNIX_LIKE) && !defined(DARWIN) && !defined(ARM32)
+#   elif defined(UNIX_LIKE) && !defined(DARWIN) && !defined(ARM32) && !(defined(SPARC) && !defined(__arch64__))
       /* Older versions of Darwin seem to lack getcontext(). */
       /* ARM Linux often doesn't support a real getcontext(). */
       ucontext_t ctxt;
